internal/rbd/controllerserver.go

```diff
@@ -318,12 +316,38 @@ func (cs *ControllerServer) CreateVolume(
 	return buildCreateVolumeResponse(req, rbdVol), nil
 }
 
-func flattenParentImage(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
+// flattenParentImage is to be called before proceeding with creating volume,
+// with datasource. This function flattens the parent image accordingly to
+// make sure no flattening is required during or after the new volume creation.
+// For parent volume, it's parent(temp clone or snapshot) is flattened.
+// For parent snapshot, the snapshot itself is flattened.
+func flattenParentImage(
+	ctx context.Context,
+	rbdVol *rbdVolume,
+	rbdSnap *rbdSnapshot,
+	cr *util.Credentials) error {
 	if rbdVol != nil {
-		// flatten the image or its parent before the reservation to avoid
+		// flatten the image's parent before the reservation to avoid
 		// stale entries in post creation if we return ABORT error and the
 		// delete volume is not called
-		err := rbdVol.flattenCloneImage(ctx)
+		// reducing the limit for cloned images to make sure the limit is in range,
+		// If the intermediate clone reaches the depth we may need to return ABORT
+		// error message as it need to be flatten before continuing, this may leak
+		// omap entries and stale temporary snapshots in corner cases, if we reduce
+		// the limit and check for the depth of the parent image clain itself we
+		// can flatten the parent images before used to avoid the stale omap entries.
+		hardLimit := rbdHardMaxCloneDepth
+		softLimit := rbdSoftMaxCloneDepth
```

Comment:
declare this outside of if statements as  its common for both snapshot and volume